glean-core/src/metrics/labeled.rs

```diff
@@ -0,0 +1,105 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashSet;
+use std::marker::PhantomData;
+
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use crate::metrics::MetricType;
+use crate::CommonMetricData;
+use crate::Glean;
+
+const MAX_LABELS: usize = 16;
+const OTHER_LABEL: &str = "__other__";
+const MAX_LABEL_LENGTH: usize = 30;
+lazy_static! {
+    static ref LABEL_REGEX: Regex = Regex::new("^[a-z_][a-z0-9_]{0,29}$").unwrap();
+}
+
+#[derive(Debug)]
+pub struct LabeledMetric<T> {
+    meta: CommonMetricData,
+    labels: Option<Vec<String>>,
+    /// Type of the underlying metric
+    /// We don't hold on to any of these, but we need to be able to create them.
+    submetric: PhantomData<T>,
+
+    seen_labels: HashSet<String>,
+}
+
+impl<T> LabeledMetric<T>
+where
+    T: MetricType,
+{
+    pub fn new(meta: CommonMetricData, labels: Option<Vec<String>>) -> LabeledMetric<T> {
+        LabeledMetric {
+            meta,
+            labels,
+            submetric: PhantomData,
+            seen_labels: HashSet::new(),
+        }
+    }
+
+    fn new_metric_with_name(&self, name: String) -> T {
+        let meta = CommonMetricData {
+            name,
+            ..self.meta.clone()
+        };
+
+        T::with_meta(meta)
+    }
+
+    fn gen_label(&mut self, label: &str) -> String {
+        format!("{}/{}", self.meta.name, label)
+    }
+
+    fn static_label<'a>(&mut self, label: &'a str) -> &'a str {
+        let labels = self.labels.as_ref().unwrap();
+        if labels.iter().any(|l| l == label) {
+            label
+        } else {
+            OTHER_LABEL
+        }
+    }
+
+    fn dynamic_label<'a>(&mut self, label: &'a str) -> &'a str {
+        // TODO: Fetch seen_labels from the database if empty
```

Comment:
Will this be taken as a follow-up or in this PR?

Reply:
Yes, [bug 1554970](https://bugzilla.mozilla.org/show_bug.cgi?id=1554970)